File: src/lib/fab.ts
import { execFile as execFileCb } from "node:child_process";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { promisify } from "node:util";
import { TMUX_TIMEOUT } from "./types";

const execFile = promisify(execFileCb);

/** Get fab progress line for a worktree. Returns null if no active change. */
export async function getStatus(worktreePath: string): Promise<string | null> {
  try {
    const { stdout } = await execFile(
      "bash",
      [
        join(worktreePath, "fab/.kit/scripts/lib/statusman.sh"),
        "progress-line",
        // statusman reads fab/current internally
      ],
      { timeout: TMUX_TIMEOUT, cwd: worktreePath },
    );
    return stdout.trim() || null;
  } catch {
    return null;
  }
}

/** Get the current active change name for a worktree. Returns null if none. */
export async function getCurrentChange(
  worktreePath: string,
): Promise<string | null> {
  try {
    const content = await readFile(
      join(worktreePath, "fab/current"),
      "utf-8",
    );
    return content.trim() || null;
  } catch {
    return null;
  }
}

/** List all changes in a worktree. Returns raw changeman output. */
export async function listChanges(
  worktreePath: string,
): Promise<string | null> {
  try {
    const { stdout } = await execFile(
      "bash",
      [join(worktreePath, "fab/.kit/scripts/lib/changeman.sh"), "list"],
      { timeout: TMUX_TIMEOUT, cwd: worktreePath },
    );
    return stdout.trim() || null;
  } catch {
    return null;
  }
}

File: packages/api/internal/fab/fab.go
package fab

import (
	"context"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

const tmuxTimeout = 10 * time.Second

// GetStatus returns the fab progress line for a worktree. Returns "" if no active change.
func GetStatus(worktreePath string) string {
	ctx, cancel := context.WithTimeout(context.Background(), tmuxTimeout)
	defer cancel()

	scriptPath := filepath.Join(worktreePath, "fab/.kit/scripts/lib/statusman.sh")
	cmd := exec.CommandContext(ctx, "bash", scriptPath, "progress-line")
	cmd.Dir = worktreePath

	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// GetCurrentChange returns the current active change name for a worktree. Returns "" if none.
func GetCurrentChange(worktreePath string) string {
	data, err := os.ReadFile(filepath.Join(worktreePath, "fab/current"))
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(data))
}

// ListChanges returns raw changeman output for a worktree.
func ListChanges(worktreePath string) string {
	ctx, cancel := context.WithTimeout(context.Background(), tmuxTimeout)
	defer cancel()

	scriptPath := filepath.Join(worktreePath, "fab/.kit/scripts/lib/changeman.sh")
	cmd := exec.CommandContext(ctx, "bash", scriptPath, "list")
	cmd.Dir = worktreePath

	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}
